File: module/python/erp.py
# ERP host function helpers.
# Parse the JSON string returned by any host fn into a typed dataclass.

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Optional

# ── Standard envelope: {success, data, message} ───────────────────────────────
# Used by: read-entity, list-entities, write-entity, delete-entity, emit-event,
#          http-request, get-time, random-uuid, cache-get, cache-set, cache-delete

@dataclass
class HostResult:
    success: bool
    data:    Optional[Any]
    message: str

def parse_result(raw: str) -> HostResult:
    try:
        v = json.loads(raw)
        return HostResult(
            success=bool(v.get("success", False)),
            data=v.get("data"),
            message=str(v.get("message", "")),
        )
    except json.JSONDecodeError as e:
        return HostResult(success=False, data=None, message=f"invalid json: {e}")

# ── Query envelope: {rows} or {error} ────────────────────────────────────────
# Used by: query

@dataclass
class QueryResult:
    rows:  list[dict[str, Any]] = field(default_factory=list)
    error: Optional[str]        = None

def parse_query(raw: str) -> QueryResult:
    try:
        v = json.loads(raw)
        if "error" in v:
            return QueryResult(error=str(v["error"]))
        return QueryResult(rows=v.get("rows", []))
    except json.JSONDecodeError as e:
        return QueryResult(error=f"parse error: {e}")

# ── Execute envelope: {affected} or {error} ───────────────────────────────────
# Used by: execute

@dataclass
class ExecuteResult:
    affected: int          = 0
    error:    Optional[str] = None

def parse_execute(raw: str) -> ExecuteResult:
    try:
        v = json.loads(raw)
        if "error" in v:
            return ExecuteResult(error=str(v["error"]))
        return ExecuteResult(affected=int(v.get("affected", 0)))
    except json.JSONDecodeError as e:
        return ExecuteResult(error=f"parse error: {e}")

# ── Response builders ─────────────────────────────────────────────────────────

def ok(data: Any = None) -> str:
    return json.dumps({"success": True, "data": data, "message": ""})

def err(message: str) -> str:
    return json.dumps({"success": False, "data": None, "message": message})

File: module/rust/src/erp.rs
// ERP host function helpers.
// Parse the JSON string returned by any host fn into a typed struct.

use serde_json::Value;

// ── Standard envelope: {success, data, message} ───────────────────────────────
// Used by: read-entity, list-entities, write-entity, delete-entity, emit-event,
//          http-request, get-time, random-uuid, cache-get, cache-set, cache-delete

pub struct HostResult {
    pub success: bool,
    pub data:    Option<Value>,
    pub message: String,
}

pub fn parse_result(raw: &str) -> HostResult {
    let v: Value = serde_json::from_str(raw).unwrap_or_default();
    HostResult {
        success: v["success"].as_bool().unwrap_or(false),
        data:    v.get("data").cloned().and_then(|d| if d.is_null() { None } else { Some(d) }),
        message: v["message"].as_str().unwrap_or("").to_string(),
    }
}

// ── Query envelope: {rows} or {error} ────────────────────────────────────────
// Used by: query

pub struct QueryResult {
    pub rows:  Vec<Value>,
    pub error: Option<String>,
}

pub fn parse_query(raw: &str) -> QueryResult {
    let v: Value = serde_json::from_str(raw).unwrap_or_default();
    if let Some(e) = v.get("error") {
        return QueryResult { rows: vec![], error: Some(e.as_str().unwrap_or("unknown error").to_string()) };
    }
    QueryResult {
        rows:  v["rows"].as_array().cloned().unwrap_or_default(),
        error: None,
    }
}

// ── Execute envelope: {affected} or {error} ───────────────────────────────────
// Used by: execute

pub struct ExecuteResult {
    pub affected: i64,
    pub error:    Option<String>,
}

pub fn parse_execute(raw: &str) -> ExecuteResult {
    let v: Value = serde_json::from_str(raw).unwrap_or_default();
    if let Some(e) = v.get("error") {
        return ExecuteResult { affected: 0, error: Some(e.as_str().unwrap_or("unknown error").to_string()) };
    }
    ExecuteResult {
        affected: v["affected"].as_i64().unwrap_or(0),
        error:    None,
    }
}

// ── Response builders ─────────────────────────────────────────────────────────

pub fn ok(data: Value) -> String {
    serde_json::json!({ "success": true, "data": data, "message": "" }).to_string()
}

pub fn ok_null() -> String {
    r#"{"success":true,"data":null,"message":""}"#.to_string()
}

pub fn err(msg: &str) -> String {
    serde_json::json!({ "success": false, "data": null, "message": msg }).to_string()
}
